apps/opik-backend/src/main/java/com/comet/opik/domain/SpanService.java

```diff
@@ -244,4 +250,58 @@ public Mono<Boolean> validateSpanWorkspace(@NonNull String workspaceId, @NonNull
         return spanDAO.getSpanWorkspace(spanIds)
                 .map(spanWorkspace -> spanWorkspace.stream().allMatch(span -> workspaceId.equals(span.workspaceId())));
     }
+
+    @Trace(dispatcher = true)
+    public Mono<Void> create(@NonNull SpanBatch batch) {
+
+        if (batch.spans().isEmpty()) {
+            return Mono.empty();
+        }
+
+        List<String> projectNames = batch.spans()
+                .stream()
+                .map(Span::projectName)
+                .distinct()
+                .toList();
+
+        Mono<List<Span>> resolveProjects = Flux.fromIterable(projectNames)
+                .flatMap(this::resolveProject)
+                .collectList()
+                .map(projects -> bindSpanToProjectAndId(batch, projects))
+                .subscribeOn(Schedulers.boundedElastic());
+
+        return resolveProjects
+                .flatMap(spanDAO::batchInsert)
+                .then();
+    }
+
+    private List<Span> bindSpanToProjectAndId(SpanBatch batch, List<Project> projects) {
+        Map<String, Project> projectPerName = projects.stream()
+                .collect(Collectors.toMap(Project::name, Function.identity()));
+
+        return batch.spans()
+                .stream()
+                .map(span -> {
+                    String projectName = WorkspaceUtils.getProjectName(span.projectName());
+                    Project project = projectPerName.get(projectName);
+
+                    if (project == null) {
+                        throw new EntityAlreadyExistsException(new ErrorMessage(List.of("Project not found")));
+                    }
+
+                    UUID id = span.id() == null ? idGenerator.generateId() : span.id();
+                    IdGenerator.validateVersion(id, SPAN_KEY);
+
+                    return span.toBuilder().id(id).projectId(project.id()).build();
+                })
+                .toList();
+    }
+
+    private Mono<Project> resolveProject(String projectName) {
```

Comment:
I believe a method already exists in this service for this. Let's avoid the duplication if possible.

Reply:
This method is already reusing the create logic. We check if it's empty

Reply:
Minor: looking at the latest code, this extra method isn't doing much , you can clean it up and just directly call 
`getOrCreateProject(WorkspaceUtils.getProjectName(projectName))`